File: virtual_cat.cs
// virtual_cat.cs
using System;
using System.IO;
using System.Text.Json;
using System.Threading;
using System.Runtime.InteropServices;

class VirtualCat
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "magenta" => "\x1b[95m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    const string CAT_ASCII = @"
 /\_/\
( o.o )
 > ^ <
";

    class State
    {
        public string Name { get; set; }
        public int Hunger { get; set; }
        public int Happiness { get; set; }
        public int Energy { get; set; }
        public int Health { get; set; }
        public int Age { get; set; }
        public string LastUpdate { get; set; }
    }

    static string ConfigFile => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".virtual_cat.json");

    static State LoadState()
    {
        if (!File.Exists(ConfigFile))
            return new State { Name = "Барсик", Hunger = 50, Happiness = 70, Energy = 80, Health = 90, Age = 0, LastUpdate = DateTime.Now.ToString("o") };
        string json = File.ReadAllText(ConfigFile);
        return JsonSerializer.Deserialize<State>(json) ?? new State();
    }

    static void SaveState(State state)
    {
        string json = JsonSerializer.Serialize(state, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(ConfigFile, json);
    }

    static string NowISO() => DateTime.Now.ToString("o");

    static int Clamp(int val, int min, int max) => Math.Max(min, Math.Min(val, max));

    static string GetEmotion(State state)
    {
        if (state.Happiness >= 80) return "😺";
        if (state.Happiness >= 50) return "😸";
        if (state.Happiness >= 30) return "😼";
        return "😿";
    }

    static string GetMoodText(State state)
    {
        if (state.Happiness >= 80) return "Мурлычет и трётся об ноги";
        if (state.Happiness >= 50) return "Играет с клубком ниток";
        if (state.Happiness >= 30) return "Сидит на подоконнике и смотрит в окно";
        return "Лежит в углу и грустит";
    }

    static void ShowStatus(State state)
    {
        Console.WriteLine(Colorize(CAT_ASCII, "magenta"));
        Console.WriteLine(Colorize($"  Имя: {state.Name}", "blue"));
        Console.WriteLine(Colorize($"  Возраст: {state.Age} лет", "blue"));
        Console.WriteLine($"  {GetEmotion(state)}  {GetMoodText(state)}");
        Console.WriteLine(Colorize($"  Голод: {state.Hunger}/100", state.Hunger > 70 ? "yellow" : "green"));
        Console.WriteLine(Colorize($"  Счастье: {state.Happiness}/100", state.Happiness > 50 ? "green" : "red"));
        Console.WriteLine(Colorize($"  Энергия: {state.Energy}/100", state.Energy > 50 ? "green" : "yellow"));
        Console.WriteLine(Colorize($"  Здоровье: {state.Health}/100", state.Health < 30 ? "red" : "green"));
    }

    static void RandomEvent(string msg = "")
    {
        string[] events = { "Мяу!", "Котик трётся о ноги.", "Принёс игрушку.",
                           "Пытается поймать муху.", "Свернулся клубком." };
        string text = string.IsNullOrEmpty(msg) ? events[new Random().Next(events.Length)] : msg;
        Console.WriteLine(Colorize($"  ✨ {text}", "yellow"));
    }

    static void AutoMode()
    {
        Console.WriteLine(Colorize("🤖  Автоматический режим включён.", "magenta"));
        var rand = new Random();
        while (true)
        {
            var state = LoadState();
            state.Hunger = Clamp(state.Hunger + rand.Next(-5, 16), 0, 100);
            state.Happiness = Clamp(state.Happiness + rand.Next(-5, 11), 0, 100);
            state.Energy = Clamp(state.Energy + rand.Next(-5, 11), 0, 100);
            state.Health = Clamp(state.Health + rand.Next(-3, 6), 0, 100);
            SaveState(state);
            Console.Clear();
            ShowStatus(state);
            Console.WriteLine(Colorize("\nНажмите Ctrl+C для выхода", "yellow"));
            Thread.Sleep(10000);
        }
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine(Colorize("Usage: virtual_cat <status|feed|play|sleep|heal|rename|auto> [name]", "yellow"));
            return;
        }
        string action = args[0];
        var state = LoadState();

        switch (action)
        {
            case "status":
                ShowStatus(state);
                break;
            case "feed":
                state.Hunger = Clamp(state.Hunger - 30, 0, 100);
                state.Happiness = Clamp(state.Happiness + 10, 0, 100);
                state.Health = Clamp(state.Health + 5, 0, 100);
                Console.WriteLine(Colorize("🐟  Ням-ням! Котик поел.", "green"));
                RandomEvent("Мурлычет от удовольствия");
                SaveState(state);
                break;
            case "play":
                if (state.Energy < 20)
                {
                    Console.WriteLine(Colorize("😿  Котик слишком устал для игр.", "red"));
                    return;
                }
                state.Happiness = Clamp(state.Happiness + 25, 0, 100);
                state.Energy = Clamp(state.Energy - 20, 0, 100);
                state.Hunger = Clamp(state.Hunger + 10, 0, 100);
                Console.WriteLine(Colorize("🧶  Игра с клубком! Котик доволен.", "green"));
                RandomEvent("Прыгает за лазерной указкой");
                SaveState(state);
                break;
            case "sleep":
                state.Energy = Clamp(state.Energy + 40, 0, 100);
                state.Hunger = Clamp(state.Hunger + 10, 0, 100);
                Console.WriteLine(Colorize("😴  Котик уснул. Сладких снов!", "blue"));
                RandomEvent("Во сне дёргает лапками");
                SaveState(state);
                break;
            case "heal":
                if (state.Energy < 20)
                {
                    Console.WriteLine(Colorize("😿  Нет сил лечить котика.", "red"));
                    return;
                }
                state.Health = Clamp(state.Health + 30, 0, 100);
                state.Energy = Clamp(state.Energy - 20, 0, 100);
                Console.WriteLine(Colorize("💊  Котик вылечен! Он благодарен.", "green"));
                SaveState(state);
                break;
            case "rename":
                if (args.Length < 2)
                {
                    Console.WriteLine(Colorize("Укажите имя: rename <имя>", "red"));
                    return;
                }
                state.Name = args[1];
                Console.WriteLine(Colorize($"🐱  Котика теперь зовут {state.Name}!", "blue"));
                SaveState(state);
                break;
            case "auto":
                AutoMode();
                break;
            default:
                Console.WriteLine(Colorize("Неизвестное действие.", "red"));
                break;
        }
    }
}

File: virtual_cat.java
// virtual_cat.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import com.google.gson.*;

public class virtual_cat {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String MAGENTA = "\u001B[95m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static final String CAT_ASCII = 
        "\n /\\_/\\\n( o.o )\n > ^ <\n";

    private static class State {
        String name;
        int hunger;
        int happiness;
        int energy;
        int health;
        int age;
        String lastUpdate;
    }

    private static String configFile = System.getProperty("user.home") + "/.virtual_cat.json";

    private static State loadState() throws IOException {
        Path path = Paths.get(configFile);
        if (!Files.exists(path)) {
            State s = new State();
            s.name = "Барсик";
            s.hunger = 50;
            s.happiness = 70;
            s.energy = 80;
            s.health = 90;
            s.age = 0;
            s.lastUpdate = Instant.now().toString();
            return s;
        }
        String json = new String(Files.readAllBytes(path));
        Gson gson = new Gson();
        return gson.fromJson(json, State.class);
    }

    private static void saveState(State state) throws IOException {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(state);
        Files.write(Paths.get(configFile), json.getBytes());
    }

    private static int clamp(int val, int min, int max) {
        return Math.max(min, Math.min(val, max));
    }

    private static String getEmotion(State state) {
        if (state.happiness >= 80) return "😺";
        if (state.happiness >= 50) return "😸";
        if (state.happiness >= 30) return "😼";
        return "😿";
    }

    private static String getMoodText(State state) {
        if (state.happiness >= 80) return "Мурлычет и трётся об ноги";
        if (state.happiness >= 50) return "Играет с клубком ниток";
        if (state.happiness >= 30) return "Сидит на подоконнике и смотрит в окно";
        return "Лежит в углу и грустит";
    }

    private static void showStatus(State state) {
        System.out.println(colorize(CAT_ASCII, MAGENTA));
        System.out.println(colorize("  Имя: " + state.name, BLUE));
        System.out.println(colorize("  Возраст: " + state.age + " лет", BLUE));
        System.out.println("  " + getEmotion(state) + "  " + getMoodText(state));
        System.out.println(colorize("  Голод: " + state.hunger + "/100",
                                    state.hunger > 70 ? YELLOW : GREEN));
        System.out.println(colorize("  Счастье: " + state.happiness + "/100",
                                    state.happiness > 50 ? GREEN : RED));
        System.out.println(colorize("  Энергия: " + state.energy + "/100",
                                    state.energy > 50 ? GREEN : YELLOW));
        System.out.println(colorize("  Здоровье: " + state.health + "/100",
                                    state.health < 30 ? RED : GREEN));
    }

    private static void randomEvent(String msg) {
        String[] events = {"Мяу!", "Котик трётся о ноги.", "Принёс игрушку.",
                           "Пытается поймать муху.", "Свернулся клубком."};
        String text = (msg == null || msg.isEmpty()) ? events[new Random().nextInt(events.length)] : msg;
        System.out.println(colorize("  ✨ " + text, YELLOW));
    }

    private static void autoMode() throws IOException, InterruptedException {
        System.out.println(colorize("🤖  Автоматический режим включён.", MAGENTA));
        Random rand = new Random();
        while (true) {
            State state = loadState();
            state.hunger = clamp(state.hunger + rand.nextInt(21) - 5, 0, 100);
            state.happiness = clamp(state.happiness + rand.nextInt(16) - 5, 0, 100);
            state.energy = clamp(state.energy + rand.nextInt(16) - 5, 0, 100);
            state.health = clamp(state.health + rand.nextInt(9) - 3, 0, 100);
            saveState(state);
            System.out.print("\033[H\033[2J");
            System.out.flush();
            showStatus(state);
            System.out.println(colorize("\nНажмите Ctrl+C для выхода", YELLOW));
            Thread.sleep(10000);
        }
    }

    public static void main(String[] args) throws IOException, InterruptedException {
        if (args.length < 1) {
            System.out.println(colorize("Usage: virtual_cat <status|feed|play|sleep|heal|rename|auto> [name]", YELLOW));
            return;
        }
        String action = args[0];
        State state = loadState();

        switch (action) {
            case "status":
                showStatus(state);
                break;
            case "feed":
                state.hunger = clamp(state.hunger - 30, 0, 100);
                state.happiness = clamp(state.happiness + 10, 0, 100);
                state.health = clamp(state.health + 5, 0, 100);
                System.out.println(colorize("🐟  Ням-ням! Котик поел.", GREEN));
                randomEvent("Мурлычет от удовольствия");
                saveState(state);
                break;
            case "play":
                if (state.energy < 20) {
                    System.out.println(colorize("😿  Котик слишком устал для игр.", RED));
                    return;
                }
                state.happiness = clamp(state.happiness + 25, 0, 100);
                state.energy = clamp(state.energy - 20, 0, 100);
                state.hunger = clamp(state.hunger + 10, 0, 100);
                System.out.println(colorize("🧶  Игра с клубком! Котик доволен.", GREEN));
                randomEvent("Прыгает за лазерной указкой");
                saveState(state);
                break;
            case "sleep":
                state.energy = clamp(state.energy + 40, 0, 100);
                state.hunger = clamp(state.hunger + 10, 0, 100);
                System.out.println(colorize("😴  Котик уснул. Сладких снов!", BLUE));
                randomEvent("Во сне дёргает лапками");
                saveState(state);
                break;
            case "heal":
                if (state.energy < 20) {
                    System.out.println(colorize("😿  Нет сил лечить котика.", RED));
                    return;
                }
                state.health = clamp(state.health + 30, 0, 100);
                state.energy = clamp(state.energy - 20, 0, 100);
                System.out.println(colorize("💊  Котик вылечен! Он благодарен.", GREEN));
                saveState(state);
                break;
            case "rename":
                if (args.length < 2) {
                    System.out.println(colorize("Укажите имя: rename <имя>", RED));
                    return;
                }
                state.name = args[1];
                System.out.println(colorize("🐱  Котика теперь зовут " + state.name + "!", BLUE));
                saveState(state);
                break;
            case "auto":
                autoMode();
                break;
            default:
                System.out.println(colorize("Неизвестное действие.", RED));
        }
    }
}
